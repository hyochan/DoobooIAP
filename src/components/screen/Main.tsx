import {
  Alert,
  EmitterSubscription,
  Image,
  Platform,
  SectionList,
  Text,
  View,
} from 'react-native';
import RNIap, {
  InAppPurchase,
  Product,
  PurchaseError,
  Subscription,
  SubscriptionPurchase,
  finishTransaction,
  purchaseErrorListener,
  purchaseUpdatedListener,
} from 'react-native-iap';
import React, { ReactElement, useCallback, useEffect, useState } from 'react';

import { Button } from '@dooboo-ui/native';
import { IC_COIN } from '../../utils/Icons';
import { RootStackNavigationProps } from '../navigation/RootStackNavigator';
import { colors } from '../../theme';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/native-theme';

const Container = styled.SafeAreaView`
  flex: 1;
  align-self: stretch;
  overflow: scroll;
  background-color: ${({ theme }): string => theme.background};

  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
  overflow: hidden;
`;

const Header = styled.View`
  height: 160px;
  width: 100%;
  background-color: ${colors.blue};
  justify-content: center;
  align-items: center;
`;

const itemSkus = Platform.select({
  default: [
    'com.dooboolab.nonconsumable',
    'com.dooboolab.consumable1000',
    'com.dooboolab.consumable5000',
    'com.dooboolab.consumable10000',
  ],
  android: [
    'android.test.purchased',
    'android.test.canceled',
    'android.test.refunded',
    'android.test.item_unavailable',
    // 'point_1000', '5000_point', // dooboolab
  ],
});

const itemSubs = Platform.select({
  default: [
    'com.dooboolab.subscription1000',
    'com.dooboolab.subscription5000',
    'com.dooboolab.subscription10000',
  ],
});

interface Props {
  navigation: RootStackNavigationProps<'Main'>;
}

interface SectionProduct {
  title: string;
  data: Product[];
}

interface SectionSubscription {
  title: string;
  data: Subscription[];
}

type Section = SectionProduct | SectionSubscription;

let purchaseUpdateSubscription: EmitterSubscription;
let purchaseErrorSubscription: EmitterSubscription;

function Intro(): React.ReactElement {
  const [sections, setSections] = useState<Section[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [totalPayedAmount, setTotalPayedAmount] = useState<number>(10000);
  const { theme } = useThemeContext();

  const getProducts = useCallback(async (): Promise<void> => {
    RNIap.clearProductsIOS();

    try {
      const result = await RNIap.initConnection();
      await RNIap.consumeAllItemsAndroid();
      console.log('result', result);
    } catch (err) {
      console.warn(err.code, err.message);
    }

    purchaseUpdateSubscription = purchaseUpdatedListener(
      async (purchase: InAppPurchase | SubscriptionPurchase) => {
        const receipt = purchase.transactionReceipt;
        if (receipt) {
          try {
            const ackResult = await finishTransaction(purchase);
            console.log('ackResult', ackResult);
          } catch (ackErr) {
            console.warn('ackErr', ackErr);
          }

          Alert.alert('purchase error', JSON.stringify(receipt));
        }
      },
    );

    purchaseErrorSubscription = purchaseErrorListener(
      (error: PurchaseError) => {
        console.log('purchaseErrorListener', error);
        Alert.alert('purchase error', JSON.stringify(error));
      },
    );

    const products = await RNIap.getProducts(itemSkus);
    products.forEach((product) => {
      product.type = 'inapp';
    });
    // console.log('products', JSON.stringify(products));
    const subscriptions = await RNIap.getSubscriptions(itemSubs);
    subscriptions.forEach((subscription) => {
      subscription.type = 'subs';
    });
    console.log('subscriptions', JSON.stringify(subscriptions));
    const list = [
      {
        title: getString('ONE_TIME_PURCHASE'),
        data: products,
      },
      {
        title: getString('SUBSCRIPTION_PURCHASE'),
        data: subscriptions,
      },
    ];
    setSections(list);
    setLoading(false);
  }, [sections]);

  useEffect(() => {
    getProducts();

    return (): void => {
      if (purchaseUpdateSubscription) {
        purchaseUpdateSubscription.remove();
      }
      if (purchaseErrorSubscription) {
        purchaseErrorSubscription.remove();
      }
    };
  }, []);

  const purchase = (item: Product | Subscription): void => {
    if (item.type === 'inapp') {
      RNIap.requestPurchase(item.productId);
    } else {
      RNIap.requestSubscription(item.productId);
    }
  };

  const renderHeader = (): ReactElement => (
    <Header>
      <Text style={{
        fontSize: 28,
        color: 'white',
      }}>{totalPayedAmount.toLocaleString()}</Text>
      <Text style={{
        marginTop: 8,
        fontSize: 16,
        color: 'white',
      }}>{getString('TOTAL_PURCHASE')}</Text>
    </Header>
  );

  const renderSectionHeader = (title: string): ReactElement => (
    <View style={{
      height: 40,
      flexDirection: 'row',
      paddingHorizontal: 16,
      backgroundColor: theme.placeholder,
    }}>
      <Text
        style={{
          fontSize: 13,
          color: theme.background,
          marginTop: 12,
        }}
      >{title}</Text>
    </View>
  );

  const renderItem = (item: Product | Subscription): ReactElement => (
    <View style={{
      padding: 16,
      flexDirection: 'row',
      backgroundColor: theme.background,
      borderBottomWidth: 1,
      borderBottomColor: theme.placeholder,
    }}>
      <Image source={IC_COIN} style={{
        height: 80,
        width: 80,
      }}/>
      <View
        style={{
          flexDirection: 'column',
          marginLeft: 32,
          marginRight: 20,
        }}
      >
        <Text
          style={{
            fontSize: 16,
            color: theme.brandLight,
            marginBottom: 4,
          }}
        >{item.title}</Text>
        <Text
          style={{
            fontSize: 14,
            color: theme.font,
          }}
        >{item.productId}</Text>
        <Button
          containerStyle={{
            width: 120,
            marginTop: 16,
          }}
          style={{
            backgroundColor: theme.background,
            height: 40,
            width: 120,
            borderWidth: 1,
            borderColor: theme.focused,
            borderRadius: 8,
            paddingHorizontal: 10,
          }}
          textStyle={{
            color: theme.font,
          }}
          onPress={(): void => purchase(item)}
          text={item.localizedPrice}
        />
      </View>
    </View>
  );

  return (
    <Container>
      <SectionList
        style={{ width: '100%' }}
        ListHeaderComponent={renderHeader}
        refreshing={loading}
        onRefresh={getProducts}
        // @ts-ignore
        sections={sections}
        keyExtractor={(item, index): string => index.toString()}
        renderItem={({ item }): ReactElement => renderItem(item)}
        renderSectionHeader={({ section: { title } }): ReactElement => renderSectionHeader(title)}
      />
    </Container>
  );
}

export default Intro;
