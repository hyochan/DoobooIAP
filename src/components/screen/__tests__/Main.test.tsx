import React, { ReactElement } from 'react';
import { RenderResult, act, fireEvent, render } from '@testing-library/react-native';
import { createTestElement, createTestProps } from '../../../../test/testUtils';

import Main from '../Main';
import { ThemeType } from '@dooboo-ui/native-theme';
import renderer from 'react-test-renderer';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: ReactElement;
let testingLib: RenderResult;

describe('[Main] screen rendering test', () => {
  beforeEach(() => {
    props = createTestProps();
    component = createTestElement(<Main {...props} />);
    testingLib = render(component);
  });

  it('should render outer component and snapshot matches', () => {
    const json = renderer.create(component).toJSON();
    expect(json).toMatchSnapshot();
    expect(json).toBeTruthy();
  });

  it('should render [Dark] theme', () => {
    component = createTestElement(<Main {...props} />, ThemeType.DARK);
    testingLib = render(component);
    const { baseElement } = testingLib;
    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });
});
