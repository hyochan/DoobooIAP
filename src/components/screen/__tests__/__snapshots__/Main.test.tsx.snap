// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Main] screen rendering test should render [Dark] theme 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <SafeAreaView
      style={
        Array [
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "backgroundColor": "#151A25",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "flex-start",
            "overflow": "hidden",
          },
        ]
      }
    >
      <ScrollView
        data={Array []}
        disableVirtualization={false}
        horizontal={false}
        initialNumToRender={10}
        maxToRenderPerBatch={10}
        onEndReachedThreshold={2}
        refreshControl={
          <RefreshControl
            onRefresh={[Function]}
            refreshing={true}
          />
        }
        refreshing={true}
        scrollEventThrottle={50}
        sections={Array []}
        stickyHeaderIndices={Array []}
        stickySectionHeadersEnabled={true}
        style={
          Object {
            "width": "100%",
          }
        }
        updateCellsBatchingPeriod={50}
        windowSize={21}
      >
        <View>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#383E76",
                  "height": 160,
                  "justifyContent": "center",
                  "width": "100%",
                },
              ]
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 28,
                }
              }
            >
              10,000
            </Text>
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 16,
                  "marginTop": 8,
                }
              }
            >
              Total Purchase
            </Text>
          </View>
        </View>
      </ScrollView>
    </SafeAreaView>
  </View>
</View>
`;

exports[`[Main] screen rendering test should render outer component and snapshot matches 1`] = `
<SafeAreaView
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "stretch",
        "backgroundColor": "#ffffff",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "justifyContent": "flex-start",
        "overflow": "hidden",
      },
    ]
  }
>
  <ScrollView
    ListHeaderComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onRefresh={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={
      <RefreshControl
        onRefresh={[Function]}
        refreshing={true}
      />
    }
    refreshing={true}
    renderItem={[Function]}
    renderSectionHeader={[Function]}
    scrollEventThrottle={50}
    sections={Array []}
    stickyHeaderIndices={Array []}
    stickySectionHeadersEnabled={true}
    style={
      Object {
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    windowSize={21}
  >
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#383E76",
              "height": 160,
              "justifyContent": "center",
              "width": "100%",
            },
          ]
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 28,
            }
          }
        >
          10,000
        </Text>
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 16,
              "marginTop": 8,
            }
          }
        >
          Total Purchase
        </Text>
      </View>
    </View>
  </ScrollView>
</SafeAreaView>
`;
